smoothscroll.js

```diff
@@ -1,156 +1,237 @@
+
+  /*
+   * returns actual time
+   * @method now
+   * @returns {Date}
+   */
   function now() {
-    return window.performance !== undefined && window.performance.now !== undefined ? window.performance.now() : Date.now !== undefined ? Date.now() : new Date().getTime();
+    return w.performance !== undefined && w.performance.now !== undefined ?
+      // if performance object supported return now, if not fallback to date object
+      w.performance.now() : Date.now !== undefined ? Date.now() : new Date().getTime();
   }
 
-  // ease-in-out
+  /*
+   * returns result of applying ease math function to a number
+   * @method ease
+   * @param {Number} k
+   * @returns {Number}
+   */
   function ease(k) {
     return 0.5 * (1 - Math.cos(Math.PI * k));
   }
 
-  function shouldBailOut(x) {
-    return typeof x === 'undefined' || x.behavior !== 'smooth';
+  /*
+   * returns true if first argument is an options object and contains a smooth behavior
+   * @method bailOut
+   * @param {Number|Object} x
+   * @returns {Boolean}
+   */
+  function bailOut(x) {
+    if (typeof x !== 'object' || x.behavior === undefined || x.behavior === 'auto' ) {
+      // first arg not an object, or behavior is auto or undefined
+      return true;
+    } else if (x.behavior === 'smooth') {
+      // first argument is an object and behavior is smooth
+      return false;
+    } else {
+      // behavior not supported
+      throw new TypeError(x.behavior + ' is not a valid value for enumeration ScrollBehavior');
+    }
+  }
+
+  /*
+   * changes scroll position inside an element
+   * @method scrollElement
+   * @params {Node} el
+   * @params {Number} x
+   * @params {Number} y
+   */
+  function scrollElement(el, x, y) {
+    el.scrollTop = y;
+    el.scrollLeft = x;
+  }
+
+  /*
+   * finds scrolalble parent of an element
+   * @method findScrollableParent
+   * @params {Node} el
+   */
+  function findScrollableParent(el) {
+    if (el.clientHeight < el.scrollHeight ||
+        el.clientWidth < el.scrollWidth) {
+      return el;
+    }
+
+    // only continue scaling if parentNode is valid
+    if (el.parentNode.parentNode) {
+      return findScrollableParent(el.parentNode);
+    }
   }
 
+  /*
+   * scrolls window with a smooth behavior
+   * @method smoothScroll
+   * @params {Number} x
+   * @params {Number} y
+   */
   function smoothScroll(x, y) {
-    var sx = window.pageXOffset;
-    var sy = window.pageYOffset;
+    var sx = w.scrollX || w.pageXOffset,
+        sy = w.scrollY || w.pageYOffset,
+        startTime = now();
 
-    if (typeof startX === 'undefined') {
+    if (startX === undefined) {
       startX = sx;
       startY = sy;
       endX = x;
       endY = y;
     }
 
-    var startTime = now();
+    // cancel frame is there is an scroll event happening
+    if (frame) {
+      w.cancelAnimationFrame(frame);
+    }
+
+    frame = w.requestAnimationFrame(step);
 
-    // TODO: look into polyfilling scroll-behavior: smooth css property
```

Comment:
Actually, by that point we could easily support it when Element.scrollBy support is added by checking for that behavior when called. No need to traverse the DOM looking for all of those elements. JIT property checks.